File: backend/src/modules/settings/settings.service.ts
import { Setting } from '@database/index';
import { parsePagination } from '@utils/pagination';
import { writeAuditLog } from '@utils/audit';
import { Op, WhereOptions } from 'sequelize';

export async function listSettings(filters: { search?: string; page?: string; limit?: string }) {
  const { page, limit, offset } = parsePagination({ ...filters, sort: 'key', order: 'asc' });

  const where: WhereOptions<Setting> = {};
  if (filters.search) {
    where.key = { [Op.like]: `%${filters.search}%` };
  }

  const { rows, count } = await Setting.findAndCountAll({
    where,
    limit,
    offset,
    order: [['key', 'ASC']],
    distinct: true,
  });

  const all = rows.map((s) => ({ [s.key]: s.value }));
  const flat = Object.assign({}, ...all);

  return {
    settings: flat,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

export async function updateSettings(actorId: string, updates: Record<string, unknown>) {
  const entries = Object.entries(updates);
  for (const [key, value] of entries) {
    const [setting] = await Setting.findOrCreate({
      where: { key },
      defaults: { key, value },
    });
    if (setting.value !== value) {
      setting.value = value;
      await setting.save();
    }
  }

  await writeAuditLog({
    actorId: Number(actorId),
    action: 'SETTINGS_UPDATED',
    module: 'setting',
    newValues: updates,
  });

  const updated = await Setting.findAll({ where: { key: { [Op.in]: entries.map(([k]) => k) } } });
  const flat = Object.assign({}, ...updated.map((s) => ({ [s.key]: s.value })));
  return flat;
}

File: backend/app/services/settings_service.py
from app.core.audit import write_audit
from app.models import Setting


def list_settings(db, filters):
    try:
        page = max(int(str(filters.get('page') or '1')), 1)
    except (TypeError, ValueError):
        page = 1
    try:
        limit = max(min(int(str(filters.get('limit') or '20')), 100), 1)
    except (TypeError, ValueError):
        limit = 20
    offset = (page - 1) * limit

    q = db.query(Setting)
    if filters.get('search'):
        q = q.filter(Setting.key.like('%{}%'.format(filters['search'])))

    total = q.count()
    rows = q.order_by(Setting.key.asc()).offset(offset).limit(limit).all()

    flat = {}
    for s in rows:
        flat[s.key] = s.value

    return {
        'settings': flat,
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }


def update_settings(db, actor_id, updates):
    for key, value in updates.items():
        setting = db.query(Setting).filter(Setting.key == key).first()
        if setting is None:
            setting = Setting(key=key, value=value)
            db.add(setting)
        elif setting.value != value:
            setting.value = value
        db.flush()

    write_audit(db, actor_id, 'SETTINGS_UPDATED', 'setting', new=updates)
    db.commit()

    keys = list(updates.keys())
    rows = db.query(Setting).filter(Setting.key.in_(keys)).all()
    flat = {}
    for s in rows:
        flat[s.key] = s.value
    return flat